### ISSUE TITLE
Removing multiple newlines from EOF?

### ISSUE THREAD

**Issue by acrewdson** [hashicorp/hcl#285]
Hello,

When using `terraform fmt` I noticed that multiple newlines are not removed (_ie_ collapsed down to a single newline) from ends of files. Would maintainers be open to a PR that would fix this?

Thanks!

**Comment by acrewdson** [comment]
@apparentlymart I PR'd this change to hcl1 before realizing that Terraform uses hcl2 🤦‍♂ https://github.com/hashicorp/hcl/pull/351/files#diff-56a7a1546fe93b5da71f86b7ae456b12R64

I don't suppose you'd be open to adding something comparably ugly (but pragmatic?) over here https://github.com/hashicorp/hcl/blob/hcl2/hclwrite/public.go#L43 ?

**Comment by Xulei-NL** [comment]
> I had a quick look at this myself today as part of some other work, and remembered why this is more complex than it first appears: the formatter is currently implemented as an in-place mutation of a list of tokens, with the goal of minimizing new allocations so that the formatter doesn't create a lot of memory pressure.
> 
> Unfortunately that means the current design doesn't currently allow the token sequence to _shrink_ during formatting, which would be required in order to remove the newline tokens.
> 
> That doesn't mean we can't do this, but it does mean we'll need to do some internal refactoring to do it, and will want to try to design it so that as much as possible it is just truncating the existing list in-place, and ideally that it's very cheap to run this extra pass over a token sequence that is _already_ formatted.

Dear Martin,

Well this Issue is four years old.

I looked at this Issue and I'm thinking of replacing [[]formatLine](https://github.com/hashicorp/hcl/blob/main/hclwrite/format.go#L27) with a linked list data structure to support adding/removing lines. Linked lists are more efficient in this case. After that, I will implement the request requested by the OP.

I found this Issue after I tried to implement formatting multl-line lists in `terraform fmt`. This is a nice first step to me. Do you think my idea is good to start off?

Best regards,

xulei

**Comment by apparentlymart** [comment]
Hi @Xulei-NL,

I don't work at HashiCorp anymore and so I'm no longer a maintainer of this codebase. I can't say whether or not HashiCorp would be open to this change, and if so whether they'd rather have it implemented here in HCL or as part of the Terraform codebase.

I would suggest that you wait to hear from someone who maintains this codebase before you do any work. My previous comments about what might be desirable are no longer valid.

**Comment by acrewdson** [comment]
thanks @apparentlymart and @Xulei-NL 

> I would suggest that you wait to hear from someone who maintains this codebase before you do any work.

@apparentlymart any idea who that would be?

**Comment by apparentlymart** [comment]
No, but I expect that you'll be able to recognize them by the "Member" badge that GitHub will put at the top right of any comments they post here.

**Comment by crw** [comment]
I will raise this for review in triage. It will be a couple of weeks for a response due to how things are currently scheduled. Thanks!

**Comment by acrewdson** [comment]
@acrewdson thanks!

**Comment by crw** [comment]
Hi @acrewdson, results from triage are that if we decide to address vertical white space in the `fmt` tool, we would prefer to examine the issue more holistically than to only trim the trailing newlines. As it stands, this is something that can be fairly easily accomplished with IDE settings or other command line based whitespace trim tools. Thanks for the submission, I will update the PRs as well.

**Comment by Xulei-NL** [comment]
> Hi [@acrewdson](https://github.com/acrewdson), results from triage are that if we decide to address vertical white space in the `fmt` tool, we would prefer to examine the issue more holistically than to only trim the trailing newlines. As it stands, this is something that can be fairly easily accomplished with IDE settings or other command line based whitespace trim tools. Thanks for the submission!

nice. I'm wondering when other formatter functions such as multi-line list formatter will be available? Is it on your radar?

for instance,

```hcl
test = [ a,
  b,
  c]
```
--->
```hcl
test = [
  a,
  b,
  c
]
```

**Comment by crw** [comment]
Not currently but we appreciate the feedback!

**Comment by Xulei-NL** [comment]
> Not currently but we appreciate the feedback!

Thanks for your reply.

May I open a separate issue to talk about it? My original motivation was to implement multi-line formatters to pretty-print list variables. 

I checked the current `terraform fmt` implementation and found it was a bit hard to have this feature due to fixed lines in formatting passes.
